procurement_auto_create_group_by_product/models/stock_rule.py

```diff
@@ -13,6 +13,10 @@ def _get_auto_procurement_group(self, product):
         if self.auto_create_group_by_product:
             if product.auto_create_procurement_group_ids:
                 return fields.first(product.auto_create_procurement_group_ids)
+            else:
+                # Make sure that two transactions can not create a procurement group
+                # For the same product at the same time.
+                self.env.cr.execute("LOCK TABLE procurement_group NOWAIT;")
```

Comment:
I think we shouldn't block the whole table. 
```suggestion
            # Make sure that two transactions can not create a procurement group
            # For the same product at the same time.
            import hashlib
            import struct
            lock_name = f"product.product,{product.id}-auto-proc-group"
            hasher = hashlib.sha1(str(lock_name).encode())
            int_lock = struct.unpack("q", hasher.digest()[:8])
            self.env.cr.execute("SELECT pg_try_advisory_xact_lock(%s);", (int_lock,))
            lock_acquired = self.env.cr.fetchone()[0]
            if not lock_acquired:
                raise UserError(f"The auto procurement group for product {product.name} is already created by someone else")
```

Reply:
Never used advisory lock. Nice thanks for that.
I also added a test for this change.